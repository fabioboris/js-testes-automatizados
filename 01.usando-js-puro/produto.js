export class Produto {
  constructor(id, descricao, preco, validade) {
    this.id = parseInt(id, 10);
    this.descricao = descricao;
    this.preco = parseFloat(preco);
    this.validade = validade;

    if (isNaN(this.id)
      || isNaN(this.preco)
      || !(this.validade instanceof Date)
      || isNaN(this.validade)) {

      throw new TypeError();
    }

    if (this.descricao.length === 0 || this.preco < 0) {
      throw new RangeError();
    }
  }

  vencido() {
    const hoje = new Date();
    hoje.setHours(0, 0, 0, 0);

    return this.validade < hoje;
  }
}