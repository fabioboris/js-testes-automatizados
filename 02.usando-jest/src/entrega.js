import { ENTREGA_URL } from '../config/entrega.config';

export async function calcularEntrega() {
  const response = await fetch(ENTREGA_URL);
  const text = await response.text();
  return parseInt(text, 10);
}