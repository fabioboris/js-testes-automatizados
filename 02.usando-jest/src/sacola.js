const _itens = [];
const _props = {};

export function Sacola(props) {
  if (props !== undefined) {
    for (const key in props) {
      _props[key] = props[key];
    }
  }
}


Sacola.prototype.adicionarItem = function (produto, quantidade) {
  if (produto.vencido()) {
    throw new RangeError(`Produto ${produto.id} vencido em ${produto.validade.toLocaleDateString('pt-BR')}`);
  }

  let item = _itens.find(item => item.produto.id === produto.id);
  if (item !== undefined) {
    item.quantidade += quantidade;
  } else {
    _itens.push({ produto, quantidade });
  }
}

Sacola.prototype.removerItem = function (produtoId) {
  const index = _itens.findIndex(item => item.produto.id === produtoId);
  if (index >= 0) {
    _itens.splice(index, 1);
  }
}

Sacola.prototype.obterItem = function (produtoId) {
  return _itens.find(item => item.produto.id === produtoId);
}

Sacola.prototype.obterItens = function () {
  return _itens;
}

Sacola.prototype.valorTotal = async function () {
  let total = _itens
    .map(item => item.produto.preco * item.quantidade)
    .reduce((acc, cur) => acc + cur);

  let entrega = 0;
  const fn = _props['calculadoraDeEntrega'];
  if (fn !== undefined) {
    entrega = await fn();
  }

  return total + entrega;
}
