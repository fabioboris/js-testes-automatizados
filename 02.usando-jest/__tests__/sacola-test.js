import { calcularEntrega } from '../src/entrega';
import { Produto } from '../src/produto';
import { Sacola } from '../src/sacola';

jest.mock('../src/entrega');


describe('Sacola', () => {

  const sacola = new Sacola({ calculadoraDeEntrega: calcularEntrega });

  const DATA_FUTURA = new Date(2020, 4, 1);
  const DATA_PRESENTE = new Date(2020, 3, 1);
  const DATA_PASSADA = new Date(2020, 2, 1);

  const produtos = [];


  beforeEach(() => {
    calcularEntrega.mockClear();
  });

  beforeAll(() => {
    jest.spyOn(Produto.prototype, 'vencido').mockImplementation(function () {
      return this.validade < DATA_PRESENTE;
    });

    produtos.push(new Produto(1, 'Leite integral 1L', 4, DATA_FUTURA));
    produtos.push(new Produto(2, 'Leite desnatado 1L', 5, DATA_FUTURA));
    produtos.push(new Produto(3, 'Pão de forma', 8, DATA_FUTURA));
    produtos.push(new Produto(4, 'Requeijão cremoso', 6, DATA_FUTURA));
    produtos.push(new Produto(5, 'Iogurte natural', 4, DATA_PASSADA));

    calcularEntrega.mockImplementation(() => 0);
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });


  test('Adicionar o primeiro item, conferir se ele consta na sacola, e se o valor total está correto', async () => {
    sacola.adicionarItem(produtos[0], 2);
    expect(sacola.obterItens()).toMatchSnapshot();
    expect(calcularEntrega).not.toHaveBeenCalled();

    const valorEsperado = 8;
    const valorObtido = await sacola.valorTotal();
    expect(valorObtido).toBe(valorEsperado);
    expect(calcularEntrega).toHaveBeenCalled();
  });

  test('Adicionar mais itens e conferir o valor total', async () => {
    sacola.adicionarItem(produtos[1], 2);
    sacola.adicionarItem(produtos[2], 2);
    sacola.adicionarItem(produtos[3], 2);

    const valorEsperado = 46;
    const valorObtido = await sacola.valorTotal();
    expect(valorObtido).toBe(valorEsperado);
    expect(calcularEntrega).toHaveBeenCalledTimes(1);
  });

  test('Remover um item e conferir o valor total', async () => {
    sacola.removerItem(2); // Leite desnatado 1L, valor: 10
    const valorEsperado = 36;
    const valorObtido = await sacola.valorTotal();
    expect(valorObtido).toBe(valorEsperado);
  });

  test('Adicionar mais unidades de um produto existente e conferir a quantidade', () => {
    sacola.adicionarItem(produtos[0], 1);
    const quantidadeEsperada = 3;
    const quantidadeObtida = sacola.obterItem(1).quantidade;
    expect(quantidadeEsperada).toBe(quantidadeObtida);
  });

  test('Emitir um erro ao tentar inserir um produto com prazo de validade vencido', () => {
    expect(() => { sacola.adicionarItem(produtos[4], 1) }).toThrow();
  });
});
