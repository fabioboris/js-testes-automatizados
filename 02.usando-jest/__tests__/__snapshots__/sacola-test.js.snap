// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sacola Adicionar o primeiro item, conferir se ele consta na sacola, e se o valor total est√° correto 1`] = `
Array [
  Object {
    "produto": Produto {
      "descricao": "Leite integral 1L",
      "id": 1,
      "preco": 4,
      "validade": 2020-05-01T03:00:00.000Z,
    },
    "quantidade": 2,
  },
]
`;
